llvm/lib/SYCLLowerIR/MutatePrintfAddrspace.cpp

```diff
@@ -0,0 +1,218 @@
+  Function *CASPrintfFunc = getCASPrintfFunction(F);
+  SmallVector<CallReplacer, 16> CallsToReplace;
+  for (User *U : F->users()) {
+    if (!isa<CallInst>(U))
+      continue;
+    auto *CI = cast<CallInst>(U);
+    CallsToReplace.emplace_back(prepareCASArgForCall(CI));
+  }
+  for (CallReplacer &CR : CallsToReplace) {
+    CR.replaceWithFunction(CASPrintfFunc);
+    ++ReplacedCallsCount;
+  }
+  if (F->hasNUses(0))
+    FunctionsToDrop.emplace_back(F);
+}
+
+/// Get the constant addrspace version of the __spirv_ocl_printf declaration, or
+/// generate it if the IR module doesn't have it yet. Also make it variadic so
+/// that it could replace all non-variadic generic AS versions.
+Function *
+AddrspaceReplacer::getCASPrintfFunction(Function *GenericASPrintfFunc) {
+  auto *CASPrintfFuncTy =
+      FunctionType::get(GenericASPrintfFunc->getReturnType(), CASLiteralType,
+                        /*isVarArg=*/true);
+  FunctionCallee CASPrintfFunc =
+      M->getOrInsertFunction("_Z18__spirv_ocl_printfPU3AS2Kcz", CASPrintfFuncTy,
+                             GenericASPrintfFunc->getAttributes());
+  auto *Callee = cast<Function>(CASPrintfFunc.getCallee());
+  Callee->setCallingConv(CallingConv::SPIR_FUNC);
+  Callee->setDSOLocal(true);
+  return Callee;
+}
+
+/// The function's effect is similar to V->stripPointerCastsAndAliases(), but
+/// also strips load/store aliases.
+Value *stripToMemorySource(Value *V) {
+  Value *MemoryAccess = V;
+  if (auto *LI = dyn_cast<LoadInst>(MemoryAccess)) {
+    Value *LoadSource = LI->getPointerOperand();
+    auto *Store = cast<StoreInst>(*llvm::find_if(
+        LoadSource->users(), [](User *U) { return isa<StoreInst>(U); }));
+    MemoryAccess = Store->getValueOperand();
+  }
+  return MemoryAccess->stripPointerCastsAndAliases();
+}
+
+/// This key function reaches the global string used as an argument to a
+/// __spirv_ocl_printf call. It then generates a constant AS copy of that global
+/// (or gets an existing one). For the return value, the call instruction is
+/// paired with its future constant addrspace string argument (or with the
+/// wrapper function argument that accepts the global).
+CallReplacer AddrspaceReplacer::prepareCASArgForCall(CallInst *CI) {
+  auto getCASLiteral = [&](GlobalVariable *Literal) -> Constant * {
+    StringRef CASLiteralName(Literal->getName().str() + "._AS2");
+    IRBuilder<> Builder(M->getContext());
+    Constant *Res = M->getOrInsertGlobal(CASLiteralName, CASLiteralType, [&] {
+      StringRef LiteralValue;
+      getConstantStringInfo(Literal, LiteralValue);
+      GlobalVariable *GV = Builder.CreateGlobalString(
+          LiteralValue, CASLiteralName, ConstantAddrspaceID, M);
+      GV->setLinkage(GlobalValue::LinkageTypes::InternalLinkage);
+      GV->setUnnamedAddr(GlobalValue::UnnamedAddr::None);
+      return GV;
+    });
+    // TODO: Create the literal in another way to ensure correct type instead
+    Res->mutateType(CASLiteralType);
+    return Res;
+  };
```

Comment:
Self-review: after `CASLiteralType` has become a member of `AddrspaceReplacer`, there's no obvious reason to have this as a lambda instead of a separate member function.